### DISCUSSION TITLE
Logdet constraints

### DISCUSSION THREAD

**Discussion by ArjunM21** [yalmip/YALMIP#1091] [Problems using YALMIP] [answered]
Hello Johan,
I am using YALMIP to solve an LMI problem. In this case, I want to add a constraint that has the form

`Constraints = [Constraints, (log(det(Y1))) <= t];`
The objective function will be of the form,

`Objective = 0.5*(-t)`

Here Y1 and t are both sdpvar. I am getting the error "Undefined function 'le' for input arguments of type 'logdet'." 

Can you help me understand what's wrong?

**Comment by johanlofberg** [comment, marked as answer]
logdet operator (which log(det()) automatically converts to) can only be used in the objective for MAXDET programming

If you want to add that nonconvex constraint you will have to avoid the automatic conversion and write, e.g., det(Y1)<= exp(t), or [log(z) <= t, z == det(Y1)]. Either way, you're not getting around the fact that it is nonconvex, and det is an awful operator.

**Reply by johanlofberg** [reply]
...but your model makes no sense, as the optimal t would be infinity, meaning the constraint would be inactive

**Reply by ArjunM21** [reply]
Thank you. 
Actually, I am converting an MINLP to a MICP problem with LMI constraints and this is just one of the constraints. Let me try what you have suggested.

**Reply by johanlofberg** [reply]
No, you should not do what I wrote, as that was just conceptual and will lead to a horrible nonconvex model which you will have absolutely no use of. To me, it looks like you simply want to have an objective -logdet(Y1), and wrote an epigraph model of that incorrectly

**Reply by ArjunM21** [reply]
This is kind of something that I am writing

```
    x = binvar(nv,1);
    z = sdpvar(1,1);
    Q1 = diag(x);
    t = sdpvar(1,1);
    Y1 = sdpvar(dof,dof);

    Constraints = [];
    Constraints = [Constraints, sum(x)==dof];
    Constraints = [Constraints, [Y1 eye(3);eye(3) C'*Q1*C] >= 0.000001*eye(6)]; 
    Constraints = [Constraints, [log(z)<=t,z==det(Y1)]];
    
    Objective = 0.5*(-t);
```

**Reply by johanlofberg** [reply]
Don't you see how that is obviously wrong as the optimal t is infinity? As I said, just use the objective -logdet(Y1) and you are done.

**Reply by johanlofberg** [reply]
...however, maximizing logdet(Y1) will be unbounded too (as maximizing log(det(Y1)) simply means making Y1 large, which clearly renders the semidefinite constraint redundant by making Y1 arbitrarily large positive semidefinite). Minimizing log(det(Y1)) would make the semidefinite constraint active and constraining, but that is not a convex problem and can thus not be done.  Writing Y1 as R^-1*R^-1 and then performing a congruence to get R in the anti-diagonal terms and then maximizing log(det(R)) would be possible I guess, if you absolutely need to use logdet and cannot accept using trace(Y1) as objective, or something similar.

**Reply by ArjunM21** [reply]
Yeah, I thought so too after thinking about what you wrote. And the problem was unbounded as you said.
I need to use logdet, so I guess I have to perform congruence. Can you help me understand what that is exactly? I am not familiar with it.

**Reply by johanlofberg** [reply]
Let Y1 = R^-1*R^-1, multiply from left and right with [R 0;0 I] and you will arrive at [I R;R C'*Q*C] which is linear in R, and now maximize logdet(R) which is equivalent with minimizing logdet(Y1)

**Reply by ArjunM21** [reply]
Thank you but it isn't clear to me. Multiply what with [R 0;0 I]? Also, I made a mistake while explaining, where I actually have to minimise -logdet(Y1) and not logdet(Y1).

**Reply by johanlofberg** [reply]
If you want to maximize determinant, i.e. minimize -logdet(Y1) you are done as that is convex and supported. However, that makes no sense, as the optimal Y1 will be infinite.

**Reply by johanlofberg** [reply]
You conceptually multiply the semidefinite constraint with [R 0;0 I] to arrive at a new semidefinite constraint which eliminates Y1 and instead is linear in R. Minimizing logdet(Y1) (which is an objective which makes sense as it will be bounded from below as you cannot set Y1 to 0) will then be equivalent to minimize logdet(R^-1R^-1) which is equivalent to maximize logdet(R) which is convex

**Reply by ArjunM21** [reply]
Hey, so I tried it kind of. 

```
%    Y1 = R^(-1) * R^(-1);
    Y1 = inv(R)*inv(R);
Constraints = [Constraints, [R 0;0 R]*[Y1 eye(3);eye(3) C'*Q1*C]*[R 0;0 R] >= 0.000001*eye(6)];
```
I tried both ways, but here I am getting the error saying R must be double or scalar. I defined R as a sdpvar just like Y1. 

Also, as you said, -logdet is already minimizing at infinity, but I think using other constraints, it won't go till infinity. Do you have any idea regarding this??

**Reply by johanlofberg** [reply]
You perform congruence and variable change on paper, to arrive at an expression which is linear in R, optimize, and then recover Y from the optimal value of R. Congruence of [Y I;I C'QC]with [R 0;0 I] leads to [I R;R C'QC] which you then constrain and optimize over

**Reply by ArjunM21** [reply]
I tried but the code doesn't run still.

Pls see the image attached. That is roughly what I want to code. C is a known matrix.
![Code](https://user-images.githubusercontent.com/91719302/158241526-76f5b8f5-f3f7-46dd-a0a7-6663ec8b4058.jpeg)

**Reply by johanlofberg** [reply]
You are moving the goalpost. Your initial problem (after removing the obviously incorrect tries) was to minimize logdet(Y) subject to [Y I;I C'*Q*C]>=0 and other stuff, which is nonconvex, but equivalent to maximize logdet(R) subject to [I R;R C'*Q*C]>0. The model you post now is nonconvex as it simply is minimize logdet(C'*Q*C) and not possible to formulate as a maxdet problem as far as I can tell

**Reply by ArjunM21** [reply]
Oh okay. This is just a thing that I tried to formulate, without being sure if it was even possible.
Will anything change if there is a trace(C'*Q*C) term as well before the -logdet term, inside the bracket?

**Reply by johanlofberg** [reply]
No. Logdet is concave, and you are trying to minimize a term involving logdet(C'*Q*C). If you really have dimension 8 (or thereabout) just enumerate and you are done. Otherwise, skip logdet and go for some other objective

**Reply by ArjunM21** [reply]
Okay, thank you. I have one more doubt. What if I have a trace(sigma) term as well in the objective function?? How will I try to solve such a function?? The constraints remain the same.

**Reply by johanlofberg** [reply]
Minimizing tr(inv(S)) can be done by adding a new variable X and then replace objective with trace(X) and the constraint X>=inv(S) which then is linearized with schur complement. In fact, this is the whole reason for Y in your original model except that you tried to use concave function logdet(Y) instead of linear function trace(Y)

**Reply by ArjunM21** [reply]
No actually the trace, as well as the logdet term, are both present. It's I think a multi-objective function. I  solved it using just the trace function like how you explained. But the original problem has them both.

**Reply by johanlofberg** [reply]
tr(CQC)^-1) - logdet(CQC) and tr(CQC) - logdet(CQC)  are the only possibilities. You cannot add logdet(CQC) to the objective, not possible to reformulate.

Multiple terms in an objective does not mean it is multi-objective.

**Reply by ArjunM21** [reply]
tr((C'QC)^-1) - logdet((C'QC)^-1) is the function actually.

**Reply by johanlofberg** [reply]
hence not possible

**Comment by ArjunM21** [comment]
Hey, so I tried something and finally could solve the -logdet part using the congruence method that you explained. My actual objective as I mentioned is to combine a trace component and a -logdet component. Can you tell me how I can combine two optimization terms into one objective function?? Like the general method for such cases? I tried something like this, but it doesn't solve anything here.

```
    Constraints = [Constraints, [Y1 S;S' C'*Q*C] >= 0.000001*eye(6)]; 
    Constraints = [Constraints, [eye(3) R;R C'*Q*C] >= 0.000001*eye(6)];
    Constraints = [Constraints, trace(Y1)>=0];
    Constraints = [Constraints, trace(Y1)<=t];


    Objective = 0.5*(t - log(det(R)) - (n-m));
```

**Reply by johanlofberg** [reply]
trace(Y1)>=0 is trivially redundant, and the objective can just as well be written as trace(Y1)-logdet(R), no reason to introduce extra variable, scale, and add constant

I don't understand what your question is though. The model you have now uses both R and Y1 which makes no sense as R is introduced in a variable change to replace the use of Y1. Do you want to minimize trace(C'*Q*C)-logdet(C'*Q*C) or what is it that you really want to minimize (disregarding any of your incorrect variable changes, Schur complements, or extra variables introduced to reformulate various stuff since these typically are red herrings as you make them incorrectly repeatedly)

**Reply by ArjunM21** [reply]
The objective function I want to minimize is trace(sigma) - logdet(sigma), where sigma is C'QC^-1. I tried replacing Y1 with R in this as well, something like this:
`
    Objective = 0.5*(trace(R) - log(det(R)) - (n-m));` 
and then made use of value o R to find the values.

But this code doesn't optimize for some reason. I also removed all redundant constraints. I don't know if it is possible to minimize this objective function, but I was able to minimize them separately and verify the results. Any idea??

**Reply by johanlofberg** [reply]
I don't now what you mean by "doesn't optimize" but it just doesn't work as you in no sense ensure that Y = R^-1R^-1 which you implicitly assume as that is where R comes from. You cannot use both the original variable and the new variable when you employ a variable change

It's simply not convex, I don't see any way of modelling this
```
>> Q = .5:.001:10;
>> plot(Q.^-1 - log(Q.^-1))
```

**Reply by ArjunM21** [reply]
I am not using the original and the new variable now together. 
Also when plotting, isn't it supposed to be 

```
>> q = 0.5:0.001:10;
>> plot(q - log(q))
```
which is convex in nature.
Sigma is (C'QC)^-1

**Reply by johanlofberg** [reply]
You just posted a model which involved both Y1 (upper bounding sigma^-1) and R (modelling inverse factor of Y1)

and when you plot, your inverse has magically disappeared. trace(X)-logdet(X) is convex, trace(X^-1)-logdet(X^-1) is nonconvex and this is what you have (with X corresponding to C'*Q*C)

**Reply by ArjunM21** [reply]
Okay, I understood this. But in reality, the whole function is convex right? Just as I had shown above? So technically it must have a minimum value at some point. Is there any way to make this work??

**Reply by johanlofberg** [reply]
What function? trace(sigma) - logdet(sigma), where sigma is (C'QC)^-1 is nonconvex as a plot easily shows for the scalar case

**Reply by ArjunM21** [reply]
Okay, but the image below is the graph that is obtained from your code. Isn't this a convex function with a minima? But let me test a little bit more and get back to you. Thank you so much for your help.
![convex](https://user-images.githubusercontent.com/91719302/159149543-03b15824-1464-44d0-9f46-fa611022bd18.jpg)

**Reply by johanlofberg** [reply]
Being convex and having a minima is not the same thing. That function is non-convex with positive curvature to the left and negative curvature to the right

**Reply by ArjunM21** [reply]
Oh okay, now I understand. Then all I have to figure out is a way to make this function into a convex one by some constraints and substitutions and that might work? If so, do you have any tips on where to start?

**Reply by johanlofberg** [reply]
Simply use a convex objective instead of a non-convex. You appear to be married to trace(X^-1)-logdet(X^-1) for some reason without even knowing its properties

**Reply by ArjunM21** [reply]
Hey Johan, thank you for your help. I am figuring out a way to do it in my way. In the meantime can you share some documents that I can go through to understand this congruencing method? I would like to have a good understanding of it since I am using it in my code. Thank you.

**Reply by johanlofberg** [reply]
Congruence transformation is simply the linear algebra concept that A>0 is equivalent with B'*A*B >0 for non-singular B, which by clever choice of B can rewrite a semidefinite constraint, possibly including a variable change and Schur complement https://yalmip.github.io/badsdps https://arxiv.org/pdf/1903.08599.pdf

**Reply by ArjunM21** [reply]
Hey, thanks but now I have a doubt. How can I replace Y1 with R in -logdet term? Shouldn't it be inv(R)*inv(R)? I understand we are trying to minimize -logdet(Y) which when we convert in terms of R becomes -logdet(inv(R)*inv(R)). Is that right?

**Reply by johanlofberg** [reply]
minimizing -logdet(Y) would not be convex in R after using that variable change (as -logdet(Y) = -logdet(R^-1R^-1) = 2logdet R)

Using R as the decision variable would only be relevant if the objective was logdet(Y) as it then would map to minimizing -logdet(R)

**Reply by ArjunM21** [reply]
Okay. So now I'm not sure what has happened. My objective function was -logdet(inv(C'QC)), which I replaced with Y to get an LMI constraint as well. Then as you had mentioned, I blindly substituted R in the place of Y to make a new LMI constraint and the objective function that I used was -logdet(R) instead of 2logdet(R). The thing is, I am getting the right result when I used -logdet(R) as the objective function. So I guess I am missing something here??

**Reply by johanlofberg** [reply]
I have no idea where you're at or what you are doing after all the confused changes back and forth. If you have the objective logdet(C'QC) as you say now, and then introduce a variable Y>=inv(C'QC) and replace the objective with logdet Y it would not work, as making Y small forces C'QC to be large which is not what the original objective wants. If you then take the Schur complemented SDP and congruence using Y = R^-1R^-1 as previously discussed, making R large through -logdet R means you are making Y small and thus again forcing C'QC large which is against the original objective logdet C'QC

And the objective logdet(C'QC)  probably makes no sense as it tries to make that matrix singular resulting in -infinity

**Reply by ArjunM21** [reply]
No the objective is -logdet(inv(C"QC)). SO our objective must be to make C"QC small so that the objective function will be small. Since the objective function has a negative sign in it, using congruencing doesn't work as Y = inv(R)*inv(R) makes the objective function 2logdet(R), which cannot be solved. 

The code written below is what I used mistakenly, but I got the right answer in the end somehow. This is wrong, as, after congruencing, the objective function would have become 2logdet(R) without negative sign.

```
Constraints = [Constraints, [eye(3) R;R C'*Q*C] >= 0.000001*eye(6)];
Objective = 0.5*(-log(det(R)));
``` 
I think I should replace inv(C'QC) with some variable such that that variable can have a large value. In that case the ojective function would be minimised.

**Reply by johanlofberg** [reply]
-logdet X^-1 = logdet X. You are now saying C'QC>=R'R and with making R large through objective -logdet R you enforce C'QC large. Equivalently inv(C'QC)<=R^-1'R^-1 which is where R came from in the original Schur form of inv(C'QC)<=Y where you tried to make Y small thus forcing C'QC large..
